File: src/main/java/mods/eln/sim/mna/component/Bipole.java
package mods.eln.sim.mna.component;

import mods.eln.sim.mna.state.State;

public abstract class Bipole extends Component {

    public State aPin, bPin;

    public Bipole() {}

    public Bipole(State aPin, State bPin) {
        connectTo(aPin, bPin);
    }


    public Bipole connectTo(State aPin, State bPin) {
        breakConnection();

        this.aPin = aPin;
        this.bPin = bPin;

        if (aPin != null) aPin.addComponent(this);
        if (bPin != null) bPin.addComponent(this);
        return this;
    }

    public Bipole connectGhostTo(State aPin, State bPin) {
        breakConnection();

        this.aPin = aPin;
        this.bPin = bPin;
        return this;
    }

    @Override
    public void breakConnection() {
        if (aPin != null) aPin.removeComponent(this);
        if (bPin != null) bPin.removeComponent(this);
    }

    @Override
    public State[] getConnectedStates() {
        return new State[]{aPin, bPin};
    }

    public abstract double getCurrent();

    public double getVoltage() {
        return (aPin == null ? 0 : aPin.state) - (bPin == null ? 0 : bPin.state);
    }


    public String toString() {
        return "[" + aPin + " " + this.getClass().getSimpleName() + " " + bPin + "]";
    }
}

File: src/main/java/mods/eln/sim/mna/component/Bipole.kt
package mods.eln.sim.mna.component

import mods.eln.sim.mna.state.State

abstract class Bipole : Component {

    var aPin: State? = null
    var bPin: State? = null

    override fun getConnectedStates() = arrayOf<State?>(aPin, bPin)

    abstract fun getCurrent(): Double

    fun getVoltage(): Double {
        return (aPin?.state ?: 0.0) - (bPin?.state ?: 0.0)
    }

    fun getBipoleU() = getVoltage()

    constructor() {}

    constructor(name: String) {
        this.name = name
    }

    constructor(aPin: State?, bPin: State?) {
        this.name = "Bipole"
        connectTo(aPin, bPin)
    }

    constructor(name:String, aPin: State?, bPin: State?) {
        this.name = name
        connectTo(aPin, bPin)
    }

    fun connectTo(aPin: State?, bPin: State?): Bipole {
        breakConnection()

        this.aPin = aPin
        this.bPin = bPin

        aPin?.add(this)
        bPin?.add(this)
        return this
    }

    fun connectGhostTo(aPin: State, bPin: State): Bipole {
        breakConnection()

        this.aPin = aPin
        this.bPin = bPin
        return this
    }

    override fun breakConnection() {
        if (aPin != null) aPin!!.remove(this)
        if (bPin != null) bPin!!.remove(this)
    }

    override fun toString(): String {
        return "[" + aPin + " " + this.javaClass.simpleName + "_" + name + " " + bPin + "]"
    }
}
